gen/generate.go

```diff
@@ -303,6 +303,11 @@ func generateModule(
 		if err = Services(g, m.Services); err != nil {
 			return "", nil, fmt.Errorf("could not generate code for services %v", err)
 		}
+	} else {
+		// Add module even if the Thrift file does not contain a service.
+		if _, err := builder.AddModule(m.ThriftPath); err != nil {
```

Comment:
Two completely divergent code paths here doesn't seem right. There's very little in common between the if-else branches.

We currently have.

```
if services { addServices() }
else { addModules() }
```

Where `addServices()` at some point does all or part of the work done by `addModules()`.

Perhaps we should do,

```
addModules()
if services { addServices() }
```

Less divergence is cleaner, easier to reason about, and easier to test.

And if you make `addModules` add *all* modules: the current module and all its included modules, then AddService just needs to look up module IDs and never actually add them.

([compile/Module.Walk][1] will help in adding a module, its included modules, and their included modules.)

[1]: https://godoc.org/go.uber.org/thriftrw/compile#Module.Walk

Reply:
I've updated the logic to reflect these suggestions. I also want to flag a change I made in the test cases, which i'll do inline there.